Tap element2

BryceStevenWilley opened the pull request:
Found a decent solution to tapping tabbed templates, and for tapping arbitrary elements. Not urgent, so leaving this in drafts until I can iterate on it for a bit longer.

BryceStevenWilley committed:
Allows you to tap an aribtrary element with a specific ID.

Necessary to navigate the `tabbed_templates` from the ALToolbox

BryceStevenWilley committed:
Tab tab specifically, and tab by general CSS selector

BryceStevenWilley committed:
Make tap elements work more consistently

Extended the timeouts in some places, don't wait for the navigation on taping
elements, and adds another step for taping an element without specifying a waittime.

BryceStevenWilley commented:
I moved over some fixes from the accessibility branch that I had working on recently, but this should be ready for review. I'll make sure the tests are passing before pinging anyone there.

plocket commented on lib/scope.js:
```diff
@@ -244,10 +244,19 @@ module.exports = {
-  afterStep: async function afterStep(scope, options = {waitForShowIf: false, navigated: false, waitForTimeout: 0}) {
+  waitForTab: async function waitForTab(scope, tab_to_show) {
+    // Implementation linked to https://github.com/SuffolkLITLab/docassemble-ALToolbox/blob/4aa1069389d2ca9cadb586f309dd73d791f081d1/docassemble/ALToolbox/misc.py#L127
+    // I think `.$()` does not use a timeout
```
I think you're right.

BryceStevenWilley replied:
This is straight copied from the above function, I believe you wrote this, haha.

plocket commented on lib/scope.js:
```diff
@@ -244,10 +244,19 @@ module.exports = {
   },
 
-  afterStep: async function afterStep(scope, options = {waitForShowIf: false, navigated: false, waitForTimeout: 0}) {
+  waitForTab: async function waitForTab(scope, tab_to_show) {
```
I've been making the second argument an object. That does complicate factors in some way, but it may be worth talking about at some point. Could you add a description of what the expected value is?

plocket replied:
Or maybe it just could have a different name, like `tab_id`.

BryceStevenWilley replied:
So it is the tab id, but not being used to get the tab, it's the labelledby text on the tab panel (where the contents are actually shown, not the item you click to show it, which is the tab). I'll add this description, don't think there's a better name that's not extremely long and unwieldy though.

I'm going to leave the arguments as is too, I don't think it needs that many options right now.

plocket commented on lib/scope.js:
```diff
@@ -244,10 +244,19 @@ module.exports = {
     /* Common things to do after a step, including take care of errors.
     *    TODO: discuss splitting into `afterField` and `afterStep` or something. Or just change name
-    *    TODO: Update to latest cucumberjs, which as `AfterStep` */
+    *    TODO: Update to latest cucumberjs, which has `AfterStep` */
```
This [already exists in v7 of cucumber](https://github.com/cucumber/cucumber-js/blob/7.x/docs/support_files/api_reference.md), a version which we use, we just haven't made the conversion yet. That said, if I recall correctly some of this doesn't actually belong after every step (hence the various arguments) and we have to think more carefully about how to handle this functionality.

plocket replied:
Also, some of this should be done while the story table runs between pages, etc, and a story table is only one Step, so we'll have to manage that manually anyway.

Not easy stuff to know about this system. I'm not yet sure how to clarify it, but a name change would probably help.

BryceStevenWilley replied:
A lot of that is out of scope for this PR: I just corrected a typo when I saw it.

plocket commented on lib/scope.js:
```diff
@@ -244,10 +244,19 @@ module.exports = {
+    if ( tab ) { await scope.page.waitForTimeout( scope.showif_timeout * 2 ); }
+  },
+
+  afterStep: async function afterStep(scope, options = {waitForShowIf: false, waitForTab: '', navigated: false, waitForTimeout: 0}) {
```
Just noting that here too `waitForTab` might be `tab_id` or `tab_id_to_wait_for` or `tab_to_wait_for`. `waitForTab` currently sounds like a boolean to me. That said, this isn't a blocking change.

plocket commented on lib/scope.js:
```diff
@@ -297,25 +307,34 @@ module.exports = {
-        // and other convos around it. We haven't made it public that the device
-        // can customized. Until we do that, we'll just use "click" all the time.
+    let inner;
+    // Error loads page, so no need to detect to keep things short
+    // Click with no navigation will end immediately
+    // Can't use elem[ scope.activate ](). It works locally, but errors on GitHub
+    // with "Node is not visible or not an HTMLElement". We're not using a custom Chromium version
+    // Unfortunately, this won't catch an invisible element. Hopefully we'd catch it before now...?
+    // "Tap" on mobile is more complex to implement ourselves.
+    // See https://stackoverflow.com/a/56547605/14144258
+    // and other convos around it. We haven't made it public that the device
+    // can customized. Until we do that, we'll just use "click" all the time.
```
Is this a good place for these comments? I'm not sure the previous place was the best place for all of these, but I think it may have been slightly better.

BryceStevenWilley replied:
It's in almost the same place? It was inside the promise.all below, but I had to duplicate that call, so it can't (and shouldn't) stay in the exact same place because it applies both to lines 321 and 326.

plocket replied:
It's a tad complex, so we can probably leave it as is for now. Here are thoughts to maybe make into an issue for future clean up.

Line 311: Think it goes under Promise.race, since it explains (I think?) why we're not racing to detect an invalid input error.

Line 312: Probably goes with the Promise that clicks and navigates.

Lines 313-319: not sure where they should go, so we can leave them here, but they should go somewhere else someday if we can keep an eye out for that. Maybe in our own repo's docs.

BryceStevenWilley replied:
I moved those first two lines to be in the right places. I think the other lines are fine where they are

plocket commented on lib/scope.js:
```diff
@@ -297,25 +307,34 @@ module.exports = {
-        // See https://stackoverflow.com/a/56547605/14144258
-        // and other convos around it. We haven't made it public that the device
-        // can customized. Until we do that, we'll just use "click" all the time.
+    let inner;
```
Just personal confusion experience - what is the name `inner` meant to describe?

BryceStevenWilley replied:
It was a lazy name: it was the inner promise of the promise race below. I've changed it to be a bit more clear.

plocket replied:
I think maybe you didn't push that change.

plocket replied:
Page hadn't updated, looks good to me!

plocket commented on lib/scope.js:
```diff
@@ -420,6 +443,16 @@ module.exports = {
+    await scope.tapElement( scope, elem, isTab=true );
+  },
+
+  tapElementById: async function ( scope, selector, waitForTimeout ) {
```
Might change this name to `tabElementBySelector` since it really can be any selector, right? It would also let people know that they have to include `#` before the selector if they want to us an id - this isn't going to add it automatically.

BryceStevenWilley replied:
This function name isn't something that authors will see, but I like the new name anyway. Changed.

plocket commented on lib/steps.js:
```diff
@@ -390,7 +390,7 @@ When(/^(?:the user gets)?(?:I get)? to ?(?:the)?(?: ?question)?(?: id)? "([^"]+)
 //#####################################
 
-When(/I wait (\d*\.?\d+) seconds?/, { timeout: -1 }, async ( seconds ) => {
+When(/I wait ?(?:for)? (\d*\.?\d+) seconds?/, { timeout: -1 }, async ( seconds ) => {
```
I like it!

plocket commented on lib/steps.js:
```diff
@@ -706,6 +706,27 @@ When('I tap to continue', { timeout: -1 }, async () => {
+
+When(/I tap the "([^"]+)" element$/i, { timeout: -1 }, async ( elemSelector ) => {
+  return wrapPromiseWithTimeout(
+    scope.tapElementById( scope, elemSelector, 2000 ),
```
Why have a hard-coded timeout? Does it make sense to use the timeout the user can set themselves? That is, `scope.timeout`?

BryceStevenWilley replied:
That's a separately defined step (that's just a different interface to this same code) that lets someone specify a timeout. IMO I didn't want to specify the timeout for every single step that I used this tap feature in, and unfortunately we do still need the time out. It's for browser events, so I can't imagine it being too flaky, since it's not based on server events. Ideally, the author / running shouldn't have to know that this timeout exists, so I don't want to make it super configurable elsewhere.

I looked into possibly using the [transition to end](https://stackoverflow.com/questions/51547994/wait-for-transition-to-end-in-puppeteer) value, but I couldn't actually get it to work for the one example I had (tapping on tabs in the tabbed interface), and I think what the actual transition that we should wait for will vary and I can't actually find?

plocket replied:
I'm willing to go with that choice.

Regarding catching transition events, you can go down the rabbit hole if you want, but it depends on the implementation being used and, yes, it would be specific to the situation. Also, tapping some things may not cause transitions, etc. It's pretty complicated even before you hit the difficulties in the code itself.

BryceStevenWilley replied:
I ran into some problems when testing this to make sure it worked, and ended up trying the transition events and it seems to work pretty well? I want to see if it'll work for the accessibility testing before merging (I remembered that the main reason I needed the wait was because I would tap the button to switch the tabs and then check for accessibility, but the page wouldn't have fully transitioned yet, so the accessibility check would check the colors with the in-progress fade-in. Those colors didn't have enough contrast, but also weren't the true colors of the page and so would be a flaky failure), but it's pretty smooth right now.

BryceStevenWilley replied:
Accessibility seems to work fine with the transition end stuff as well. I think it should be good.

plocket replied:
I'll have to see the code to see if it's general enough. To use it here, this would have to be able to detect the transition of tapping any element, not just a tab. If you want to just detect tab tapping transitions, that should probably be a separate Step just for tabs.

plocket replied:
Somehow my page wasn't updating with new code. I'll check it out now and see how it all flows.

plocket commented on lib/scope.js:
```diff
@@ -297,25 +307,34 @@ module.exports = {
+      // Same thing as above, but the promise doesn't wait for navigation.
+      inner = Promise.all([
+        elem.evaluate( (el) => {return el.click(); }),
+      ]);
```
Just a style note - `.evaluate()` returns a promise, so you shouldn't need to use `Promise.all()` here.

plocket reviewed (approved):
Great addition! None of these comments are blocking changes, so I'll approve, but maybe we can make some into issues. I think the two main concerns that I noted are:

First, the hard-coded timeout in the `tap "..." element`. Even though there's a way to have a custom timeout, I still don't see why we'd have a hidden hard-coded timeout.

Second, the name of `tapElementById`, which I think is a misleading name.

Thanks for all the work!

BryceStevenWilley committed:
Add tap test to interactive_steps.feature

Uses the "test_taps" in ALAutomatedTestingTests, clicks through several
different tabs made by the `tabbed_templates_html` ALToolbox function and a
url action button.

BryceStevenWilley committed:
Add transitionend event listener

Much more consistent, and still pretty fast.
Also corrected some poor variable names, from suggestions in PR review.

BryceStevenWilley commented:
Test interview file in https://github.com/plocket/docassemble-ALAutomatedTestingTests/pull/187. Tested locally and passing 100%.

BryceStevenWilley committed:
Correct misspelling of museum

BryceStevenWilley committed:
Moved comments to more relevant locations

plocket commented on tests/features/interactive_steps.feature:
```diff
@@ -101,3 +101,15 @@ Scenario: handles settings from Github secrets
+  Then I see the phrase "villify"
+  And I tap the "Tests-third_template-tab" tab
+  Then I see the phrase "museum"
+  And I tap the "#special_event" element and wait 5 seconds
```
Future issue: Would waiting 1 second be enough? I don't see a need for 5 and our tests are going to continue growing. I'd rather keep the time we add manually to a minimum.

Doesn't need to block this PR.

BryceStevenWilley replied:
I thought I had a reason for 5, but 1 second seems to work fine locally at least. I changed it to 1 (and added another fix to the step regex to allow for grammatically correct steps), hopefully it's not too flaky.

plocket commented on lib/scope.js:
```diff
@@ -420,13 +459,23 @@ module.exports = {
     // If there is no browser open, start a new one
     if (!scope.browser) {
-      scope.browser = await scope.driver.launch({ headless: !session_vars.get_debug() });
+      scope.browser = await scope.driver.launch({ headless: !session_vars.get_debug(), devtools: session_vars.get_debug() });
```
Nice find!

BryceStevenWilley replied:
Thanks! I don't think it's necessary as a comment, but I found it from trying to debug some of the puppeteer `evaluate` functions: https://stackoverflow.com/a/49887254/11416267.

plocket replied:
Just under that I see a link to https://github.com/puppeteer/puppeteer#debugging-tips with more tips.

BryceStevenWilley committed:
Shorten the tap wait in the feature test

Also allows for singular "second" on the "tap the _ element and wait X seconds"
step.

plocket commented:
Future: I think one remaining improvement we could make in the future is to better handle when the tab doesn't exist. For example, for when the user has a typo.

plocket commented:
> How you you suggest handling it?

I think I'd try to detect the tab that has the actual ID. If it exists, can we assume its content exists too?

plocket commented:
I suggest not using `waitForSelector` as that will slow everything down. Unless there's a reason you can think that the element would take a while to appear on the page.

BryceStevenWilley committed:
Catch if the user mispells tab/element selector

Fails and prints the message "Couldn't find ... on the page, is there a typo?"

BryceStevenWilley committed:
Added information about tap steps to the CHANGELOG

BryceStevenWilley commented:
Everything should be good (addressed the last comment, and added info to the changelog). The documentation is ready in https://github.com/SuffolkLITLab/docassemble-AssemblyLine-documentation/pull/244 and will be merged soon. I'll squash and merge now.
